src/components/autocomplete/AutocompleteConnected.tsx

```diff
@@ -0,0 +1,173 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import * as _ from 'underscore';
+import {IReactVaporState} from '../../ReactVapor';
+import {keyCode} from '../../utils/InputUtils';
+import {IDispatch, ReduxConnect} from '../../utils/ReduxUtils';
+import {IItemBoxProps} from '../itemBox/ItemBox';
+import {ListBoxConnected} from '../listBox/ListBoxConnected';
+import {addAutocomplete, removeAutocomplete, setAutocompleteActive, setAutocompleteValue, toggleAutocomplete} from './AutocompleteActions';
+import {IAutocompleteState} from './AutocompleteReducers';
+
+const defaultMatchFilter = (filterValue: string, item: IItemBoxProps) => {
+    if (filterValue === '') {
+        return true;
+    }
+
+    const regex = new RegExp(filterValue, 'gi');
+    return regex.test(item.value) || regex.test(item.displayValue);
+};
+
+export interface IAutocompleteOwnProps {
+    id: string;
+    items: IItemBoxProps[];
+    matchFilter?: (filterValue: string, item: IItemBoxProps) => boolean;
+}
+
+export interface IAutocompleteStateProps {
+    isOpen?: boolean;
+    value?: string;
+    visibleItems?: IItemBoxProps[];
+    active?: number;
+}
+
+export interface IAutocompleteDispatchProps {
+    onRender?: () => void;
+    onDestroy?: () => void;
+    onDocumentClick?: () => void;
+    onFocus?: () => void;
+    onChange?: (value: string) => void;
+    setActive?: (diff: number) => void;
+}
+
+export interface IAutocompleteProps extends IAutocompleteOwnProps, IAutocompleteStateProps, IAutocompleteDispatchProps {}
+
+const mapStateToProps = (state: IReactVaporState, ownProps: IAutocompleteOwnProps): IAutocompleteStateProps => {
+    const autocomplete: IAutocompleteState = _.findWhere(state.autocompletes, {id: ownProps.id});
+    const listbox = _.findWhere(state.listBoxes, {id: ownProps.id});
+    const defaultValue = listbox && listbox.selected && listbox.selected.length ? listbox.selected[0] : '';
+    const value = autocomplete && autocomplete.value || defaultValue;
+
+    const itemsWithHidden = _.map(ownProps.items, (item: IItemBoxProps): IItemBoxProps => {
+        const visible = _.isUndefined(ownProps.matchFilter)
+            ? defaultMatchFilter(value, item)
+            : ownProps.matchFilter(value, item);
+
+        return {...item, hidden: !visible || !!item.hidden};
+    });
+
+    let index = 0;
+    const activeIndex = autocomplete && autocomplete.active;
+    const visibleLength = _.filter(itemsWithHidden, (item: IItemBoxProps) => !item.hidden && !item.disabled).length;
+    const mod = (x: number, n: number) => (x % n + n) % n; // mod is a modulo function that works with negative numbers
+    const visibleItems = _.map(itemsWithHidden, (item: IItemBoxProps): IItemBoxProps => {
+        let active = false;
+        if (!item.hidden && !item.disabled) {
+            active = mod(activeIndex, visibleLength) === index;
+            index++;
+        }
+        return {...item, active};
+    });

```

Comment:
I don't think this (and `itemsWithHidden`) should be here. It is not part of the state. It should be done in the component itself.

Reply:
I'm 100% against moving the logic, but I think it makes more sense to put it here since it re-use the pattern of the single select component